### PR TITLE
Allow server handlers to send response headers directly

### PR DESCRIPTION
**Pull request by glbrntt**
Motivation:

The async server call context allows users to set headers which are sent when the first message is sent. In many cases this is fine, however, some use cases require the headers to be sent immediately.

Modifications:

- Add `sendHeaders(_:)` to the `GRPCAsyncServerCallContext` which sends headers to the client and throws if headers have already been written or it's too late to send them.

Result:

Headers can be sent directly from a server call handler.

### CODE REVIEW

File: Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift [hunk 1]
```diff
@@ -747,11 +760,44 @@ extension AsyncServerHandler: AsyncServerCallContextProvider {
...
+        case .cancel:
+          return self.cancel(error: nil)
+        case .drop:
+          ()
+        }
+
+      case .intercept(.none), .drop:
+        // intercept(.none) means headers have already been sent; we should never hit this because
+        // we guard on setting the response headers above.
+        ()
```
Comment (FranzBusch) [inline Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift:795]: If we never should hit this should we precondition here?

File: Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerCallContext.swift [hunk 2]
```diff
@@ -30,6 +30,21 @@ public struct GRPCAsyncServerCallContext: Sendable {
...
+  /// On accepting the RPC the server will send the given headers (which may be empty) along with
+  /// any transport specific headers (such the ":status" pseudo header) to the client.
+  ///
+  /// It is not necessary to call this function: the RPC is implicitly accepted when the first
+  /// response message is sent, however this may be useful when clients require an early indication
+  /// that the RPC has been accepted.
+  ///
+  /// If the RPC has already been accepted (either implicitly or explicitly) then this function is
+  /// a no-op.
+  public func acceptRPC(headers: HPACKHeaders) async {
```
Comment (FranzBusch) [inline Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerCallContext.swift:44]: What happens if I call `setHeaders` after using `acceptRPC`? Is `setHeaders` going to throw or just silently fail?

**Review by FranzBusch — COMMENTED** [review]
Overall LGTM. Just two smaller comments

File: Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerCallContext.swift (hunk 2 above)
Comment (glbrntt) [reply Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerCallContext.swift:44]: It'll throw.

File: Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift (hunk 1 above)
Comment (glbrntt) [reply Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift:795]: I think it's a bit aggressive here: there's no harm in us reaching and ignoring this state.

Comment (FranzBusch) [reply Sources/GRPC/AsyncAwaitSupport/GRPCAsyncServerHandler.swift:795]: Should we call this out in the docs?

**Review by FranzBusch — APPROVED** [review]
LGTM!

**Review by Lukasa — APPROVED** [review]

**Pull request merged**